Integrate FidryCpuCoreCounter

theofidry opened the pull request:
See https://github.com/theofidry/cpu-core-counter/issues/11 for the motivations.

The code should be sensibly the same, the notable differences are:

- fixed the deprecated usage of `hw.ncpu`
- /proc/cpuinfo is check _last_
- nproc is checked first (before was not checked at all, it's considered to be more accurate and less convoluted than cpuinfo though)
- not sure about the `return 2`, but this was not too clear [here](https://github.com/phpstan/phpstan-src/pull/514#pullrequestreview-660926363) neither. I could otherwise change the fallback value to return `1` if `proc_open` doesn't exist and 2 otherwise

I would also wait on 1.x before merging this if you're happy with it. Technically the current version is fine, I'm just looking for a round of feedback before publishing 1.0.

ondrejmirtes commented:
One idea I have - it should differentiate between physical and virtual cores, offer different getters for that. That'd be really valuable. I want to use the virtual cores, no matter the operating system.

I understand that on systems with hyperthreading, `virtual cores = 2 * physical cores`. The CPU wouldn't be fully utilized if I used only the physical cores.

But of course it might be more complicated, I'm not a hardware expert.

ondrejmirtes commented:
And if PHPStan should be able to use this library, right now it needs to support PHP 7.2+.

theofidry committed:
Improve the CPU core count detection

See https://github.com/theofidry/cpu-core-counter/issues/11 for the motivations.

The code should be sensibly the same, the notable differences are:

- fixed the deprecated usage of `hw.ncpu`
- /proc/cpuinfo is check _last_
- nproc is checked first (before was not checked at all, it's considered to be more accurate and less convoluted than cpuinfo though)
- not sure about the `return 2`, but this was not too clear [here](https://github.com/phpstan/phpstan-src/pull/514#pullrequestreview-660926363) neither. I could otherwise change the fallback value to return `1` if `proc_open` doesn't exist and 2 otherwise
- add more ways to find the CPU cores count

theofidry commented:
@ondrejmirtes should be good; there is now a distinction possible between physical and logical cores. The default picked is logical (as PHP source does and as was done previously here).

If you fancy checking out what finder finds what, to find a bug or play around with the order or something, I can recommend [those commands](https://github.com/theofidry/cpu-core-counter#checks-what-finders-find-what-on-your-system)

ondrejmirtes commented:
I really like your work, the library looks nice!

But it looks like there's a related failure: https://github.com/phpstan/phpstan-src/actions/runs/3647730351/jobs/6160310175

ondrejmirtes commented on src/Process/CpuCoreCounter.php:
```diff
@@ -24,42 +16,11 @@ public function getNumberOfCpuCores(): int
-		if (!function_exists('proc_open')) {
-			return $this->count = 1;
+		try {
+			return (new FidryCpuCoreCounter())->getCount();
```
also the caching into `$this->count` should probably be kept here :)

theofidry replied:
Missed that 🤦

theofidry commented:
Indeed and the failure is related I think. It's curious though as this is a piece of code coming from Infection and I thought I tested the case where the commands would fail... will look into it, but will probably have to rely on the CI here for that, at least partially

theofidry commented:
Nevermind it's windows related, I'll be able to check that on a machine at home soon :)

theofidry commented:
@ondrejmirtes green

ondrejmirtes commented:
Not really:

* https://github.com/phpstan/phpstan-src/actions/runs/3648030637/jobs/6161060710
* https://github.com/phpstan/phpstan-src/actions/runs/3648030605/jobs/6160966865 (still the same problem with Windows)

theofidry commented:
@ondrejmirtes FINALLY it is green

ondrejmirtes commented:
Nice :) Please definitely be ready for a round of feedback once this is released, maybe we broke some scenario that isn't in the test suite, but that's fine, it's an "unknown unknown", so if we learn about it, the better :)

ondrejmirtes commented:
Thank you!
